pkg/cmd/agent-task/capi/sessions.go

```diff
@@ -0,0 +1,208 @@
+		}
+
+		q := req.URL.Query()
+		q.Set("page_size", strconv.Itoa(perPage))
+		q.Set("page_number", strconv.Itoa(page))
+		req.URL.RawQuery = q.Encode()
+
+		res, err := c.httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+		defer res.Body.Close()
+		if res.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("failed to list sessions: %s", res.Status)
+		}
+		var response struct {
+			Sessions []session `json:"sessions"`
+		}
+		if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
+			return nil, fmt.Errorf("failed to decode sessions response: %w", err)
+		}
+		if len(response.Sessions) == 0 || len(sessions) >= limit {
+			break
+		}
+		sessions = append(sessions, response.Sessions...)
+		page++
+	}
+
+	// Drop any above the limit
+	if len(sessions) > limit {
+		sessions = sessions[:limit]
+	}
+
+	// Hydrate the Sessions with pull request data.
+	Sessions, err := c.hydrateSessionPullRequests(sessions)
+	if err != nil {
+		return nil, err
+	}
+
+	return Sessions, nil
+}
+
+// hydrateSessionPullRequests hydrates pull request information in sessions
+func (c *CAPIClient) hydrateSessionPullRequests(sessions []session) ([]*Session, error) {
+	if len(sessions) == 0 {
+		return nil, nil
+	}
+
+	prNodeIds := make([]string, 0, len(sessions))
+
+	for _, session := range sessions {
+		prNodeID := generatePullRequestNodeID(int64(session.RepoID), session.ResourceID)
+		if slices.Contains(prNodeIds, prNodeID) {
+			continue
+		}
+		prNodeIds = append(prNodeIds, prNodeID)
+	}
+
+	apiClient := api.NewClientFromHTTP(c.httpClient)
+
+	var resp struct {
+		Nodes []struct {
+			PullRequest sessionPullRequest `graphql:"... on PullRequest"`
+		} `graphql:"nodes(ids: $ids)"`
+	}
+
+	host, _ := c.authCfg.DefaultHost()
+	err := apiClient.Query(host, "FetchPRs", &resp, map[string]any{
+		"ids": prNodeIds,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	prs := make([]*api.PullRequest, 0, len(prNodeIds))
+	for _, node := range resp.Nodes {
+		prs = append(prs, &api.PullRequest{
+			ID:             node.PullRequest.ID,
+			FullDatabaseID: node.PullRequest.FullDatabaseID,
+			Number:         node.PullRequest.Number,
+			Title:          node.PullRequest.Title,
+			State:          node.PullRequest.State,
+			URL:            node.PullRequest.URL,
+			Body:           node.PullRequest.Body,
+			CreatedAt:      node.PullRequest.CreatedAt,
+			UpdatedAt:      node.PullRequest.UpdatedAt,
+			ClosedAt:       node.PullRequest.ClosedAt,
+			MergedAt:       node.PullRequest.MergedAt,
+			Repository:     node.PullRequest.Repository,
+		})
+	}
+
+	newSessions := make([]*Session, 0, len(sessions))
+	// For each session, we need to attach the Pull Request
+	for _, s := range sessions {
+		// For each Pull Request, check if it matches the session
+		for _, pr := range prs {
+			if strconv.FormatInt(s.ResourceID, 10) == pr.FullDatabaseID {
+				newSessions = append(newSessions, &Session{
+					session:     s,
+					PullRequest: pr,
+				})
+			}
+		}
+	}
```

Comment:
How about using a map to make it $O(n)$?


```suggestion
	prMap := make(map[string]*api.PullRequest, len(prNodeIds))
	for _, node := range resp.Nodes {
		prMap[node.PullRequest.FullDatabaseID] = &api.PullRequest{
			ID:             node.PullRequest.ID,
			FullDatabaseID: node.PullRequest.FullDatabaseID,
			Number:         node.PullRequest.Number,
			Title:          node.PullRequest.Title,
			State:          node.PullRequest.State,
			URL:            node.PullRequest.URL,
			Body:           node.PullRequest.Body,
			CreatedAt:      node.PullRequest.CreatedAt,
			UpdatedAt:      node.PullRequest.UpdatedAt,
			ClosedAt:       node.PullRequest.ClosedAt,
			MergedAt:       node.PullRequest.MergedAt,
			Repository:     node.PullRequest.Repository,
		}
	}

	newSessions := make([]*Session, 0, len(sessions))
	for _, s := range sessions {
		newSessions = append(newSessions, &Session{
			session:     s,
			PullRequest: prMap[strconv.FormatInt(s.ResourceID, 10)],
		})
	}
```

Reply:
Fixed in ea9dfae3bf16b6730e4d5e23e6ca5caefd8b7c68